File: input/load.fs
module input.load

open core.smap
open core.sudoku
open core.puzzlemap

(* Load a sudoku given as a single line of gridSize*gridSize characters *)
let loadPuzzle (cells : cell list) (alphabetisedLine : digit option list) : SMap<cell, digit option> = 
    List.zip cells alphabetisedLine
    |> SMap.ofList

let load (puzzleShape : puzzleShape) (sudoku : string) : solution = 

    let charToDigit (trialDigit : char) : digit option = 
        let compareAlpha (Digit charDigit) = trialDigit = charDigit
        List.tryFind compareAlpha puzzleShape.alphabet

    let alphabetisedLine =
        sudoku
        |> List.ofSeq
        |> List.map charToDigit

    let p = tPuzzleMap puzzleShape

    let given = loadPuzzle p.cells alphabetisedLine

    let current = givenToCurrent p.cells given (Digits.ofList puzzleShape.alphabet)

    { solution.given = given
      current = current
      steps = [ Load sudoku ] }

File: src/load.ml
open Sudoku
open Puzzlemap

(* Load a sudoku given as a single line of gridSize*gridSize characters *)
let loadPuzzle (cells : cells) (alphabetisedLine : digit option list) : given =
    Given (Sset.zip (Cells.toList cells) alphabetisedLine)

let load (puzzleShape : puzzleShape) (sudoku : string) : solution = 

    let charToDigit (trialDigit : char) : digit option = 
        let compareAlpha (Digit charDigit) = trialDigit = charDigit in
        let digits = Digits.toList puzzleShape.alphabet in
        if List.exists compareAlpha digits then
            Some (List.find compareAlpha digits)
        else None
        in

    let alphabetisedLine : digit option list =
        sudoku
        |> Sset.explode
        |> List.map charToDigit
        in

    let p = tPuzzleMap puzzleShape in

    let given = loadPuzzle p.cells alphabetisedLine in

    let current = Solution.givenToCurrent p.cells given puzzleShape.alphabet in

    { given = given;
      current = current;
      steps = [ Load sudoku ] }
